pkg/bootconfig/zipconfig.go

```diff
@@ -0,0 +1,130 @@
+// its signature with the provided public key file. The signature is expected to
+// be appended to the ZIP file and have fixed length `ed25519.SignatureSize` .
+// The returned string argument is the temporary directory where the files were
+// extracted, if successful.
+// No decoder (e.g. JSON, ZIP) or other function parsing the input file is called
+// before verifying the signature.
+func FromZip(filename string, pubkeyfile *string) (*Manifest, string, error) {
+	// load the whole zip file in memory - we need it anyway for the signature
+	// matching.
+	// TODO refuse to read if too big?
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, "", err
+	}
+	zipbytes := data
+	// Load the public key
+	if pubkeyfile != nil {
+		zipbytes = data[:len(data)-ed25519.SignatureSize]
+		pubkey, err := crypto.LoadPublicKeyFromFile(*pubkeyfile)
+		if err != nil {
+			return nil, "", err
+		}
+
+		// Load the signature.
+		// The signature is appended to the zip file and has length
+		// `ed25519.SignatureSize`. We read these bytes from the end of the file and
+		// treat them as the attached signature.
+		signature := data[len(data)-ed25519.SignatureSize : len(data)]
+		if len(signature) != ed25519.SignatureSize {
+			return nil, "", fmt.Errorf("Short read when reading signature: want %d bytes, got %d", ed25519.SignatureSize, len(signature))
+		}
+
+		// Verify the signature against the public key and the zip file bytes
+		if ok := ed25519.Verify(pubkey, zipbytes, signature); !ok {
+			return nil, "", fmt.Errorf("Invalid ed25519 signature for file %s", filename)
+		}
+	} else {
+		log.Printf("No public key specified, the ZIP file will be unpacked without verification")
+	}
+
+	// At this point the signature is valid. Unzip the file and decode the boot
+	// configuration.
+	r, err := zip.NewReader(&memoryZipReader{Content: zipbytes}, int64(len(zipbytes)))
+	if err != nil {
+		return nil, "", err
+	}
+	tempDir, err := ioutil.TempDir(os.TempDir(), "bootconfig")
+	if err != nil {
+		return nil, "", err
+	}
+	log.Printf("Created temporary directory %s", tempDir)
+	var manifest *Manifest
+	for _, f := range r.File {
+		destination := path.Join(tempDir, f.Name)
+		// NOTE I don't know if it's possible to have a zero-length file name in
+		// a zip file. However I am not checking for this case as I prefer this
+		// to blow up with a panic and highlight the unexpected malformed file.
+		if f.Name[len(f.Name)-1] == '/' {
+			// it's a directory, create it
+			if err := os.MkdirAll(destination, os.ModeDir|os.FileMode(0700)); err != nil {
+				return nil, "", err
+			}
+			log.Printf("Extracted directory %s (flags: %d)", f.Name, f.Flags)
+		} else {
+			fd, err := f.Open()
+			if err != nil {
+				return nil, "", err
+			}
+			buf, err := ioutil.ReadAll(fd)
+			if err != nil {
+				return nil, "", err
+			}
+			if f.Name == "manifest.json" {
+				// parse the Manifest containing the boot configurations
+				manifest, err = NewManifest(buf)
```

Comment:
So if we have multiple manifests in the zip, the latest one in the decompression order is the only one taken into account right ? Since it's not especially simple/well-known how to control file order in a zip file this might warrant a warning in documentation that you *probably* want to only have one manifest ever

Reply:
Yes - just checked that you can have duplicate files in a ZIP (and that, uh... "makes sense"), https://play.golang.org/p/kfvLvKxnzlP 

```
$ unzip -l duplicate.zip 
Archive:  duplicate.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        5  1980-00-00 00:00   duplicate
        6  1980-00-00 00:00   duplicate
---------                     -------
       11                     2 files
```

I will add a comment and a warning if a duplicate manifest is found.